physical/azure/azure_ha.go

```diff
@@ -0,0 +1,428 @@
+	defer metrics.MeasureSince(metricLockValue, time.Now())
+
+	r, err := l.get(context.Background())
+	if err != nil {
+		return false, "", err
+	}
+	if r == nil {
+		return false, "", err
+	}
+	return true, string(r.Value), nil
+}
+
+// attemptLock attempts to acquire a lock. If the given channel is closed, the
+// acquisition attempt stops. This function returns when a lock is acquired or
+// an error occurs.
+func (l *Lock) attemptLock(stopCh <-chan struct{}) (bool, error) {
+	ticker := time.NewTicker(l.retryInterval)
+	defer ticker.Stop()
+
+	for {
+		// writeLock intentionally not place within case<-ticker to avoid the
+		// initial retry interval wait
+		acquired, err := l.writeLock()
+		if err != nil {
+			return false, fmt.Errorf("attempting lock: %w", err)
+		}
+		if acquired {
+			return true, nil
+		}
+
+		select {
+		case <-ticker.C:
+		case <-stopCh:
+			return false, nil
+		}
+	}
+}
+
+// renewLock renews the given lock until the channel is closed.
+func (l *Lock) renewLock() {
+	ticker := time.NewTicker(l.renewInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			l.writeLock()
+		case <-l.stopCh:
+			return
+		}
+	}
+}
+
+// watchLock checks whether the lock has changed in the table and closes the
+// leader channel accordingly. If an error occurs during the check, watchLock
+// will retry the operation and then close the leader channel if it can't
+// succeed after retries.
+func (l *Lock) watchLock() {
+	retries := 0
+	ticker := time.NewTicker(l.watchRetryInterval)
+
+OUTER:
+	for {
+		// Check if the channel is already closed
+		select {
+		case <-l.stopCh:
+			break OUTER
+		default:
+		}
+
+		// Check if we've exceeded retries
+		if retries >= l.watchRetryMax-1 {
+			break OUTER
+		}
+
+		// Wait for the timer
+		select {
+		case <-ticker.C:
+		case <-l.stopCh:
+			break OUTER
+		}
+
+		// Attempt to read the key
+		r, err := l.get(context.Background())
+		if err != nil {
+			retries++
+			continue
+		}
+
+		// Verify the identity is the same
+		if r == nil || r.Identity != l.identity {
+			break OUTER
+		}
+	}
+
+	l.stopLock.Lock()
+	defer l.stopLock.Unlock()
+	if !l.stopped {
+		l.stopped = true
+		close(l.stopCh)
+	}
+}
+
+// writeLock writes the given lock using the following algorithm:
+//
+// - lock does not exist
+//   - write the lock
+// - lock exists
+//   - if key is empty or identity is the same or timestamp exceeds TTL
+//     - update the lock to self
+func (l *Lock) writeLock() (bool, error) {
+	l.backend.permitPool.Acquire()
+	defer l.backend.permitPool.Release()
+
+	// Create a transaction to read and the update (maybe)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// The operation may be retried, so we need to stop it if we lose leadership.
+	go func() {
+		select {
+		case <-l.stopCh:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+
+	// Build up the access conditions
+	var conds azblob.BlobAccessConditions
+
+	// Read the record
+	r, err := l.get(ctx)
```

Comment:
This seems like a race - someone could acquire a lock in between our get and our update, and then we'd blow it away.  I don't know enough Azure to offer good advice here, but I think you should try to eliminate the get if possible, instead relying solely on the conds to protect the update.

Reply:
The outcome of the get informs how the access condition should be built. The access condition influences the behavior of the subsequent write.

If the lock file doesn't already exist, the cond will be `If-None-Match: *`, which means that the file must not exist in order for the subsequent write to succeed.

If the lock file already exists, the cond will be `If-Match: {eTag_of_lock_file}`, which means that the file must exist and has not been modified for the subsequent write to succeed.

Reply:
How about instead of doing the get, you take note of the Etag in the BlockBLobUploadResponse you get back from the blob.Upload call?  Then the renews can make use of that for their IfMatch condition, and if you don't have an Etag (because this is your first writeLock call, i.e. it's not a renewal) you use IfNoneMatch:* as before?

As to the timestamp check below, maybe instead of that, use IfUnmodifiedSince in your update?

These two changes would save you a roundtrip and also eliminate my race concern.

Reply:
Thinking about it a bit more I don't think my suggestion above quite works.  Let me try again.

For the initial case, we want to grab the lock unless there's a lock that has been recently updated.  So for that initial case, maybe the only condition needed is IfUnmodifiedSince=(now-ttl).  But I don't know how IfUnmodifiedSince behaves if the blob doesn't exist, so maybe we also need If-None-match:*.  But the docs say that these conditions are ANDed together whereas we really want an OR, and I don't really understand the table they give describing the combined behaviour here: https://docs.microsoft.com/en-us/rest/api/storageservices/specifying-conditional-headers-for-blob-service-operations.

One option in the initial case if we can't find a way to express the desired condition: first try to update with If-None-Match:*, and if that fails attempt it with IfUnmodifiedSince.  Or the other way around, whichever.

For the renew case, we want to grab the lock if we still own it.  So there the only condition needed is IfMatch = (etag returned by our Upload).  We don't really care about the ttl in this case.

Reply:
Thanks for suggesting these alternate methods.

The current implementation does not have a race condition.